tests/test_utils.py

```diff
@@ -0,0 +1,45 @@
+from dataclasses import dataclass
+
+import pytest
+from itemadapter import ItemAdapter
+
+from duplicate_url_discarder.utils import item_signature, load_keys_from_path
+
+
+class Object1:
+    pass
+
+
+class Object2:
+    pass
+
+
+def test_load_keys_from_path() -> None:
+    assert load_keys_from_path({}) == {}
+
+    mapping = {"tests.test_utils.Object1": "object1", Object2: "object2"}
+    assert load_keys_from_path(mapping) == {Object1: "object1", Object2: "object2"}
+
+    with pytest.raises(
+        ValueError, match="Error loading object 'does-not-exist': not a full path"
+    ):
+        load_keys_from_path({"does-not-exist": True})
+
+
+def test_item_signature() -> None:
+    @dataclass
+    class FakeItem:
+        name: str
+
+    value = "fake_item"
+    item = FakeItem(value)
+    adapter = ItemAdapter(item)
+
+    assert item_signature(adapter, ["name"]) == hash(value)
+
+    exception_text = (
+        "Got type <class 'tests.test_utils.test_item_signature.<locals>.FakeItem'> "
+        "but expected ItemAdapter."
+    )
+    with pytest.raises(ValueError, match=exception_text):
+        item_signature(item, ["name"])  # type: ignore
```

Comment:
If we keep the current implementation, maybe we could add an xfail test for:

```
item1 = {"a": "a|", "b": "b"}
item2 = {"a": "a", "b": "|b"}
fields = ["a", "b"]
signature = "a||b"
```

Reply:
Very interesting edge case! ✨ 

Let's go with adding the attrib name _(or dict key)_ in the hash calculation. As this would mean that the left-hand value is guaranteed to be unique since it's being used as a dict-key or an attribute name. https://github.com/zytedata/duplicate-url-discarder/pull/17/commits/afd2a0f031fa33da71fa834fe0b0c61f4f87d212

Reply:
Then:

```
item1 = {"a": "a|b:b", "b": "b"}
item2 = {"a": "a", "b": "b|b:b"}
fields = ["a", "b"]
signature = "a:a|b:b|b:b"
```

:slightly_smiling_face:

Reply:
😆 

Tempted to go with the route of closer to `json.dumps()` but still faster:

```python
>>> python -m timeit "item = {'a':1,'b':2,'c':3}; import json; hash(json.dumps({k:item.get(k) for k in ['a', 'b', 'c']}))"
200000 loops, best of 5: 1.41 usec per loop

>>> python -m timeit "item = {'a':1,'b':2,'c':3}; hash('|'.join(f'{k}:{item.get(k)}' for k in ['a', 'b', 'c']))"
500000 loops, best of 5: 599 nsec per loop

>>> python -m timeit "item = {'a':1,'b':2,'c':3}; hash(','.join(f'\"{k}\": \"{item.get(k)}\"' for k in ['a', 'b', 'c']))"
500000 loops, best of 5: 649 nsec per loop
```

Reply:
I really think the original implementation was OK, it would be extremely unlikely for this to be an issue with real data, and if that ever becomes a concern, we could always allow customizing `|` into something else. But I think it makes sense to have an xfail test, as a “we know this can happen in theory” statement in code.

Reply:
Added xfail in https://github.com/zytedata/duplicate-url-discarder/pull/17/commits/c562bc6099dd7c3f2dfd8597ca20d8d161f15632. Cheers again for raising!